File: packages/core/src/format/format.ts
import { stringify as stringifyYaml } from 'yaml';
import { SECRET_NAME_PATTERN } from '../validation/secret-name';

/**
 * Rendering secrets into the file formats people export them to.
 *
 * The whole module is one property: **whatever goes in comes back out
 * byte-for-byte**. A quoting bug here does not throw, it hands someone a
 * password with a character missing, and they find out when a service fails to
 * authenticate at 3am. `format.test.ts` asserts the round-trip through this
 * repository's own parsers for a set of deliberately hostile values; that test
 * is the actual specification of the escaping rules below.
 *
 * Where a format cannot represent a value, this throws rather than emitting
 * something close. A truncated secret that looks fine is worse than an error.
 */

export type ExportFormat = 'env' | 'json' | 'yaml' | 'shell' | 'docker';

export interface ExportableSecret {
  name: string;
  value: string;
}

export class ExportFormatError extends Error {
  constructor(
    readonly secretName: string,
    reason: string,
  ) {
    // The name is safe to put in an error message; the value never is, because
    // this string will be logged.
    super(`Cannot export "${secretName}": ${reason}`);
    this.name = 'ExportFormatError';
  }
}

export function formatSecrets(
  secrets: ReadonlyArray<ExportableSecret>,
  format: ExportFormat,
): string {
  for (const secret of secrets) assertRepresentableName(secret.name, format);

  switch (format) {
    case 'env':
      return joinLines(secrets.map((s) => `${s.name}=${quoteEnvValue(s.value)}`));
    case 'shell':
      return joinLines(secrets.map((s) => `export ${s.name}=${quoteSingle(s)}`));
    case 'docker':
      return joinLines(secrets.map((s) => `${s.name}=${dockerValue(s)}`));
    case 'json':
      return `${JSON.stringify(toSortedRecord(secrets), null, 2)}\n`;
    case 'yaml':
      return stringifyYaml(toSortedRecord(secrets), {
        /**
         * Line folding is reversible in theory and a liability in practice: it
         * only round-trips if every consumer implements folding identically,
         * and the cost of one that does not is a corrupted secret. Long values
         * stay on one line.
         */
        lineWidth: 0,
        /**
         * Quote every value, always. An unquoted `yes`, `NO`, `null`, `1.10` or
         * `08:00` is a string under YAML 1.2 and something else entirely under
         * YAML 1.1 — which is what PyYAML and much of the Ruby and PHP world
         * still implement. `PASSWORD: yes` reaching Python as the boolean
         * `True` is a real failure mode, and quoting costs two characters.
         */
        defaultStringType: 'QUOTE_DOUBLE',
        /** Keys are validated secret names, so they need no quoting. */
        defaultKeyType: 'PLAIN',
      });
  }
}

/**
 * `env`, `shell` and `docker` are line-oriented and keep the caller's order —
 * which is the order the UI displayed and the user reasoned about. `json` and
 * `yaml` are documents whose key order carries no meaning, so they are sorted:
 * a secret added today should produce a one-line diff, not a reshuffled file.
 *
 * Sorting is by code unit rather than `localeCompare`, so the output does not
 * depend on the server's locale.
 */
function toSortedRecord(secrets: ReadonlyArray<ExportableSecret>): Record<string, string> {
  const sorted = [...secrets].sort((a, b) => (a.name < b.name ? -1 : a.name > b.name ? 1 : 0));
  return Object.fromEntries(sorted.map((secret) => [secret.name, secret.value]));
}

function joinLines(lines: readonly string[]): string {
  return lines.map((line) => `${line}\n`).join('');
}

/**
 * A name that is not a valid identifier would produce a line that parses back
 * as something else — `A=B=value` reads as `A` holding `B=value`. Names come
 * from the database and are already constrained, so this is a guard against a
 * future caller, not a validation step.
 */
function assertRepresentableName(name: string, format: ExportFormat): void {
  if (!SECRET_NAME_PATTERN.test(name)) {
    throw new ExportFormatError(name, `"${name}" is not a valid secret name for ${format} output.`);
  }
}

/**
 * Values that force quoting in a `.env` file.
 *
 * Whitespace and `#` change where the value ends. Quotes and backslashes change
 * how it is read. `$` and backtick do not matter to this repository's parser —
 * it never interpolates — but they do to python-dotenv, Docker Compose and
 * anything that sources the file, so a value containing one is quoted to keep
 * those readers from expanding it.
 */
const ENV_NEEDS_QUOTING = /[\s#"'$`\\]/;

/**
 * Quotes a `.env` value, and only when something in it requires quoting.
 *
 * Single quotes are preferred because they are literal in every dialect: no
 * escapes to get wrong, and no interpolation even in readers that perform it.
 * Double quotes are used only for values a single-quoted string cannot express
 * — one containing a `'`, since the literal form has no way to escape its own
 * delimiter — or a newline, which is escaped rather than written literally so
 * that one line of output stays one secret. A file truncated mid-write then
 * loses a secret instead of merging the remainder into the previous value.
 */
function quoteEnvValue(value: string): string {
  if (value === '' || !ENV_NEEDS_QUOTING.test(value)) return value;
  if (!/['\n\r]/.test(value)) return `'${value}'`;

  const escaped = value
    .replace(/\\/g, '\\\\')
    .replace(/"/g, '\\"')
    .replace(/\n/g, '\\n')
    .replace(/\r/g, '\\r')
    .replace(/\t/g, '\\t');

  return `"${escaped}"`;
}

/**
 * The only generally safe way to quote for a shell: single quotes, with the
 * `'\''` idiom for an embedded quote — close the string, emit an escaped quote,
 * reopen it. Nothing inside single quotes is expanded, so a value containing
 * `$(rm -rf …)` is inert when the file is sourced.
 *
 * A newline needs no escaping — inside single quotes it is literal — but a
 * carriage return is refused. Single quotes have no escape mechanism at all, so
 * the CR would have to be written raw, and every reader of the resulting file
 * (this repository's parsers included) normalises line endings before looking
 * at the content. The byte would be silently lost on the way back in, which is
 * the one outcome worth throwing over. `env` and `json` escape it and are the
 * formats to use for a value that contains one.
 */
function quoteSingle(secret: ExportableSecret): string {
  if (secret.value.includes('\r')) {
    throw new ExportFormatError(
      secret.name,
      'shell output cannot represent a carriage return, which would be lost when the file is read back. Export it as env or json instead.',
    );
  }
  return `'${secret.value.replace(/'/g, "'\\''")}'`;
}

/**
 * Docker's `--env-file` grammar has no quoting at all: everything after the
 * first `=` is the value, verbatim, to the end of the line.
 *
 * That makes a newline unrepresentable — Docker would read the remainder as a
 * new variable, or drop it. Emitting a value silently cut at the first newline
 * is exactly the failure this module exists to prevent, so it is an error.
 *
 * It also means a value containing ` #` or leading whitespace exports fine here
 * but is not `.env`-compatible, because that grammar is verbatim and `.env` is
 * not. That is a property of `--env-file`, and the reason `env` is the format
 * to reach for by default.
 */
function dockerValue(secret: ExportableSecret): string {
  if (/[\n\r]/.test(secret.value)) {
    throw new ExportFormatError(
      secret.name,
      'Docker env files cannot represent a value containing a line break. Export it as env, json or yaml instead.',
    );
  }
  return secret.value;
}

File: cli/cmd/xecret/format.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"regexp"
	"sort"
	"strings"
)

// Rendering secrets into the file formats people export them to — on this side
// of the wire, because for an end-to-end encrypted environment there is no other
// side that could.
//
// This is a port of `packages/core/src/format/format.ts`, and it holds the same
// single property: **whatever goes in comes back out byte for byte**. A quoting
// bug here does not throw, it hands somebody a password with a character
// missing, and they find out when a service fails to authenticate at 3am. Where
// a format cannot represent a value this refuses, because a truncated secret
// that looks fine is worse than an error.
//
// The two implementations exist because the two modes render in different
// places, not because anybody wanted two. The escaping rules are stated once, in
// that file's comments, and reproduced here; `format_test.go` asserts the
// round-trip that keeps them honest.

// envNeedsQuoting is what forces quoting in a .env file.
//
// Whitespace and `#` change where the value ends. Quotes and backslashes change
// how it is read. `$` and backtick do not matter to a parser that never
// interpolates, but they do to python-dotenv, Docker Compose, and anything that
// sources the file.
var envNeedsQuoting = regexp.MustCompile("[\\s#\"'$`\\\\]")

// formatSecrets renders a decrypted environment.
//
// `env`, `shell` and `docker` are line-oriented and keep name order; `json` and
// `yaml` are documents whose key order carries no meaning, so they are sorted so
// that a secret added today produces a one-line diff. Sorting is by code unit,
// never by locale, so the output does not depend on where it ran.
func formatSecrets(secrets map[string]string, format string) ([]byte, error) {
	names := make([]string, 0, len(secrets))
	for name := range secrets {
		// A name that is not an identifier would produce a line that parses back
		// as something else — `A=B=value` reads as `A` holding `B=value`. Names
		// come from the database and are already constrained, so this guards
		// against a future caller rather than validating one.
		if !secretNamePattern.MatchString(name) {
			return nil, fmt.Errorf("cannot export %q: it is not a valid secret name", name)
		}
		names = append(names, name)
	}
	sort.Strings(names)

	var out bytes.Buffer
	switch format {
	case "env":
		for _, name := range names {
			fmt.Fprintf(&out, "%s=%s\n", name, quoteEnvValue(secrets[name]))
		}

	case "shell":
		for _, name := range names {
			quoted, err := quoteSingle(name, secrets[name])
			if err != nil {
				return nil, err
			}
			fmt.Fprintf(&out, "export %s=%s\n", name, quoted)
		}

	case "docker":
		for _, name := range names {
			value, err := dockerValue(name, secrets[name])
			if err != nil {
				return nil, err
			}
			fmt.Fprintf(&out, "%s=%s\n", name, value)
		}

	case "json":
		encoder := json.NewEncoder(&out)
		// Matching JSON.stringify: two-space indent, and no HTML escaping, which
		// Go does by default and JavaScript does not. A value containing `<`
		// would otherwise come back as `<` — still correct JSON, still the
		// same string after parsing, but a gratuitous difference between the two
		// modes in a file people read.
		encoder.SetIndent("", "  ")
		encoder.SetEscapeHTML(false)
		if err := encoder.Encode(secrets); err != nil {
			return nil, err
		}

	case "yaml":
		for _, name := range names {
			// Every value is quoted, always. An unquoted `yes`, `NO`, `null`,
			// `1.10` or `08:00` is a string under YAML 1.2 and something else
			// entirely under YAML 1.1 — which PyYAML and much of the Ruby and PHP
			// world still implement. `PASSWORD: yes` reaching Python as the
			// boolean True is a real failure mode, and quoting costs two
			// characters. Keys are validated names and need none.
			fmt.Fprintf(&out, "%s: %s\n", name, quoteYamlValue(secrets[name]))
		}

	default:
		return nil, fmt.Errorf("unknown format %q — use env, json, yaml, shell or docker", format)
	}

	return out.Bytes(), nil
}

// quoteEnvValue quotes a .env value, and only when something requires it.
//
// Single quotes are preferred because they are literal in every dialect: no
// escapes to get wrong, and no interpolation even in readers that perform it.
// Double quotes are used only for a value single quotes cannot express — one
// containing a `'`, since the literal form cannot escape its own delimiter — or
// a newline, which is escaped rather than written literally so that one line of
// output stays one secret.
func quoteEnvValue(value string) string {
	if value == "" || !envNeedsQuoting.MatchString(value) {
		return value
	}
	if !strings.ContainsAny(value, "'\n\r") {
		return "'" + value + "'"
	}

	escaped := strings.NewReplacer(
		`\`, `\\`,
		`"`, `\"`,
		"\n", `\n`,
		"\r", `\r`,
		"\t", `\t`,
	).Replace(value)
	return `"` + escaped + `"`
}

// quoteSingle is the only generally safe way to quote for a shell: single
// quotes, with the `'\”` idiom for an embedded quote — close the string, emit
// an escaped quote, reopen it. Nothing inside single quotes is expanded, so a
// value containing `$(rm -rf …)` is inert when the file is sourced.
//
// A newline needs no escaping; a carriage return is refused. Single quotes have
// no escape mechanism, so the CR would be written raw, and every reader of the
// file normalises line endings before looking at the content — the byte would be
// silently lost on the way back in.
func quoteSingle(name, value string) (string, error) {
	if strings.Contains(value, "\r") {
		return "", fmt.Errorf(
			"cannot export %q: shell output cannot represent a carriage return, "+
				"which would be lost when the file is read back. Use --format env or json",
			name,
		)
	}
	return "'" + strings.ReplaceAll(value, "'", `'\''`) + "'", nil
}

// dockerValue: Docker's --env-file grammar has no quoting at all — everything
// after the first `=` is the value, verbatim, to the end of the line.
//
// That makes a line break unrepresentable: Docker would read the remainder as a
// new variable, or drop it. Emitting a value silently cut at the first newline
// is exactly the failure this file exists to prevent.
func dockerValue(name, value string) (string, error) {
	if strings.ContainsAny(value, "\n\r") {
		return "", fmt.Errorf(
			"cannot export %q: Docker env files cannot represent a value containing a "+
				"line break. Use --format env, json or yaml",
			name,
		)
	}
	return value, nil
}

// quoteYamlValue renders a YAML double-quoted scalar.
//
// Written out rather than delegated to a YAML library, because the property that
// matters is a scalar every parser reads back identically, and a library's
// choice of block style, folding or plain scalars is exactly what would break
// that. Line folding in particular is reversible only if every consumer
// implements it the same way, and the cost of one that does not is a corrupted
// secret — so long values stay on one line.
func quoteYamlValue(value string) string {
	var out strings.Builder
	out.WriteByte('"')
	for _, r := range value {
		switch r {
		case '\\':
			out.WriteString(`\\`)
		case '"':
			out.WriteString(`\"`)
		case '\n':
			out.WriteString(`\n`)
		case '\r':
			out.WriteString(`\r`)
		case '\t':
			out.WriteString(`\t`)
		default:
			if r < 0x20 || r == 0x7f {
				fmt.Fprintf(&out, `\x%02x`, r)
				continue
			}
			out.WriteRune(r)
		}
	}
	out.WriteByte('"')
	return out.String()
}
